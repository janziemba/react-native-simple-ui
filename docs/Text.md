# Text

## Props
Extend React Native's [Text](https://facebook.github.io/react-native/docs/text#props) props
- `color` (_string_, default null)
- `size` (_string_, default normal)

## Examples

```jsx

```
