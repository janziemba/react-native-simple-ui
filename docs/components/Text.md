---
id: text
title: Text
---
## Props
Extend React Native's [Text](https://facebook.github.io/react-native/docs/text#props) props
- `color` (_string_, default null) - one of theme colors, [base theme colors](../Theme.md#colors) by default
- `size` (_string_, default normal) - one of theme sizes, [base theme sizes](../Theme.md#sizes) by default

<img src="../assets/Text/1.png" alt="Text" width="250" style="display: inline-block;">
