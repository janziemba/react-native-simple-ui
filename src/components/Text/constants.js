// @flow

import colors from '../../themes/base/colors';
import type { ObjectOfStringsType } from '../../types';

export const COLORS: ObjectOfStringsType = {
    ...colors.palette,
    ...colors.social,
    black: 'black',
    white: 'white',
};

export const SIZES: ObjectOfStringsType = {
    h1: 'h1',
    h2: 'h2',
    h3: 'h3',
    h4: 'h4',
    h5: 'h5',
    h6: 'h6',
    normal: 'normal',
    small: 'small',
};
