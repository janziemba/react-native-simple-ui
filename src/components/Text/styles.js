// @flow

import merge from 'lodash/merge';

import { themeType } from '../../types';
import { COLORS, SIZES } from './constants';

export default ({ colors, spacing, typography }: themeType): {} => {
    const allColors: {} = merge({}, COLORS, colors.palette, colors.social);
    const colorOptions: {} = {};

    Object.keys(allColors).forEach((colorName: string) => {
        const color: string = allColors[colorName];

        colorOptions[colorName] = {
            color,
        };
    });

    return {
        base: {
            color: colors.text.primary,
            fontFamily: typography.fontFamily,
            textAlign: 'center',
        },
        colors: colorOptions,
        sizes: {
            [SIZES.h1]: {
                ...typography.sizes.h1,
                marginBottom: spacing.medium,
            },
            [SIZES.h2]: {
                ...typography.sizes.h2,
                marginBottom: spacing.medium,
            },
            [SIZES.h3]: {
                ...typography.sizes.h3,
                marginBottom: spacing.medium,
            },
            [SIZES.h4]: {
                ...typography.sizes.h4,
                marginBottom: spacing.medium,
            },
            [SIZES.h5]: {
                ...typography.sizes.h5,
                marginBottom: spacing.small,
            },
            [SIZES.h6]: {
                ...typography.sizes.h6,
                marginBottom: spacing.small,
            },
            [SIZES.normal]: {
                ...typography.sizes.normal,
            },
            [SIZES.small]: {
                ...typography.sizes.small,
            },
        },
    };
};
