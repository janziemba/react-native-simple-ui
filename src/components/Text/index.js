import PropTypes from 'prop-types';
import React, { PureComponent } from 'react';
import { Text as RNText } from 'react-native';

import withTheme from '../../themes/withTheme';
import hasStyleChanged from '../../utils/hasStyleChanged';
import { SIZES } from './constants';
import styles from './styles';

const getStyles = (props) => {
    const { color, size, styles } = props;

    return [
        styles.base,
        styles.sizes[size],
        color ? styles.colors[color] : {},
    ];
};

const propTypes = {
    ...RNText.propTypes, // eslint-disable-line react/forbid-foreign-prop-types
    color: PropTypes.string,
    size: PropTypes.oneOf(Object.keys(SIZES)),
    styles: PropTypes.objectOf(PropTypes.object).isRequired,
};

const defaultProps = {
    ...RNText.defaultProps,
    color: null,
    size: SIZES.normal,
};

class Text extends PureComponent {
    constructor(props) {
        super(props);

        this.state = {
            styles: getStyles(props),
        };
    }
    componentWillReceiveProps(nextProps) {
        const propsOnWhichDependsTheStyle = ['color', 'size'];

        if (hasStyleChanged(propsOnWhichDependsTheStyle, nextProps, this.props)) {
            this.setState({ styles: getStyles(nextProps) });
        }
    }
    render() {
        const { children } = this.props;
        const { styles } = this.state;

        return (
            <RNText
                {...this.props}
                style={styles}
            >
                {children}
            </RNText>
        );
    }
}

Text.propTypes = propTypes;
Text.defaultProps = defaultProps;

export default withTheme(styles, 'Text')(Text);
