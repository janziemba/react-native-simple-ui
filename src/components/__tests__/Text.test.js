import React from 'react';
import renderer from 'react-test-renderer';

import Text from '../Text';

it('renders correctly', () => {
    const tree = renderer.create(<Text>Lorem ipsum</Text>).toJSON();
    expect(tree).toMatchSnapshot();
});
