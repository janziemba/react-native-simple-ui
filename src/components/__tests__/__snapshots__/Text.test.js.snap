// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Text
  color={null}
  size="normal"
  style={
    Array [
      Object {
        "color": "#3d3d3d",
        "fontFamily": "System",
        "textAlign": "center",
      },
      Object {
        "fontSize": 14,
        "fontWeight": "700",
      },
      Object {},
    ]
  }
  styles={
    Object {
      "base": Object {
        "color": "#3d3d3d",
        "fontFamily": "System",
        "textAlign": "center",
      },
      "colors": Object {
        "black": Object {
          "color": "black",
        },
        "danger": Object {
          "color": "#ff4136",
        },
        "dark": Object {
          "color": "#484848",
        },
        "disabled": Object {
          "color": "#999999",
        },
        "facebook": Object {
          "color": "#3b5998",
        },
        "google": Object {
          "color": "#dc4e41",
        },
        "info": Object {
          "color": "#75caeb",
        },
        "instagram": Object {
          "color": "#c32aa3",
        },
        "light": Object {
          "color": "#f6f6f6",
        },
        "linkedin": Object {
          "color": "#0077b5",
        },
        "primary": Object {
          "color": "#158cba",
        },
        "success": Object {
          "color": "#28b62c",
        },
        "twitter": Object {
          "color": "#55acee",
        },
        "vk": Object {
          "color": "#4c75a3",
        },
        "warning": Object {
          "color": "#ff851b",
        },
        "white": Object {
          "color": "white",
        },
      },
      "sizes": Object {
        "h1": Object {
          "fontSize": 31,
          "fontWeight": "900",
          "marginBottom": 12,
        },
        "h2": Object {
          "fontSize": 25,
          "fontWeight": "900",
          "marginBottom": 12,
        },
        "h3": Object {
          "fontSize": 19,
          "fontWeight": "800",
          "marginBottom": 12,
        },
        "h4": Object {
          "fontSize": 17,
          "fontWeight": "800",
          "marginBottom": 12,
        },
        "h5": Object {
          "fontSize": 16,
          "fontWeight": "800",
          "marginBottom": 8,
        },
        "h6": Object {
          "fontSize": 15,
          "fontWeight": "700",
          "marginBottom": 8,
        },
        "normal": Object {
          "fontSize": 14,
          "fontWeight": "700",
        },
        "small": Object {
          "fontSize": 12,
        },
      },
    }
  }
  theme={
    Object {
      "colors": Object {
        "layout": Object {
          "backgroundColor": "#ffffff",
        },
        "palette": Object {
          "danger": "#ff4136",
          "dark": "#484848",
          "disabled": "#999999",
          "info": "#75caeb",
          "light": "#f6f6f6",
          "primary": "#158cba",
          "success": "#28b62c",
          "warning": "#ff851b",
        },
        "social": Object {
          "facebook": "#3b5998",
          "google": "#dc4e41",
          "instagram": "#c32aa3",
          "linkedin": "#0077b5",
          "twitter": "#55acee",
          "vk": "#4c75a3",
        },
        "text": Object {
          "disabled": "#999999",
          "placeholder": "#e2e4e6",
          "primary": "#3d3d3d",
          "secondary": "#9e9e9e",
        },
      },
      "components": Object {},
      "radiuses": Object {
        "gigantic": 16,
        "huge": 12,
        "large": 8,
        "medium": 6,
        "small": 4,
        "tiny": 2,
      },
      "spacing": Object {
        "gigantic": 24,
        "huge": 20,
        "large": 16,
        "medium": 12,
        "small": 8,
        "tiny": 4,
      },
      "typography": Object {
        "fontFamily": "System",
        "fontSizes": Object {
          "large": 16,
          "medium": 14,
          "small": 12,
        },
        "fontWeights": Object {
          "bold": "900",
          "medium": "800",
          "normal": "600",
        },
        "iconSet": "MaterialIcons",
        "sizes": Object {
          "h1": Object {
            "fontSize": 31,
            "fontWeight": "900",
          },
          "h2": Object {
            "fontSize": 25,
            "fontWeight": "900",
          },
          "h3": Object {
            "fontSize": 19,
            "fontWeight": "800",
          },
          "h4": Object {
            "fontSize": 17,
            "fontWeight": "800",
          },
          "h5": Object {
            "fontSize": 16,
            "fontWeight": "800",
          },
          "h6": Object {
            "fontSize": 15,
            "fontWeight": "700",
          },
          "normal": Object {
            "fontSize": 14,
            "fontWeight": "700",
          },
          "small": Object {
            "fontSize": 12,
          },
        },
      },
    }
  }
>
  Lorem ipsum
</Text>
`;
